### PR TITLE
Adds configurable scheduler to support mid-training

### PR DESCRIPTION
**Pull request by undfined**

### CODE REVIEW

**Comment by undfined** [comment]
Example run: https://beaker.allen.ai/orgs/ai2/workspaces/oe-data/work/01JPTSK6HV128VB27G0M9NX0C2/logs?jobId=01JPTSK6PAS23591P0ZP9A2QX4

**Comment by soldni** [comment]
This looks good to me, @dirkgr better for review.

File: src/cookbook/model/builder.py [hunk 1]
```diff
@@ -86,25 +100,25 @@ class TransformerConfigBuilder:
         get_warmup_steps() -> int:
             Returns the number of warmup steps.
 
-        get_batch_size() -> int:
-            Returns the global batch size based on the sequence length.
+        get_batch_sizes() -> tuple[int, int]:
+            Returns the global and rank microbatch sizes based on the sequence length.
```
Comment (dirkgr) [inline src/cookbook/model/builder.py:104]: The reason global batch size and microbatch size are kept separate is that the global batch size is a setting that makes a mathematical difference to how training progresses, changes training dynamics, interacts with the learning rate, etc., while the microbatch size is an implementation detail that can change depending on which hardware you use, how many GPUs you have, and how you've optimized your run.

It's not wrong per se to return them together here, but it's pretty unusual. They are very different settings.

File: src/cookbook/model/builder.py [hunk 2]
```diff
@@ -339,19 +357,24 @@ def build_dataset_config(self) -> NumpyDatasetConfig:
...
+    def get_scheduler_config(self) -> Scheduler:
+        # TODO(undfined): Move these scheduler_type strings into an enum
+        if self.scheduler_type == "cosine_warmup":
             return CosWithWarmup(warmup_steps=self.get_warmup_steps())
-        elif scheduler_type == "wsd":
+        elif self.scheduler_type == "cosine_nowup":
+            return CosWithWarmup(warmup_steps=0)
+        elif self.scheduler_type == "linear_nowup":
+            return LinearWithWarmup(warmup_steps=0, t_max=self.max_tokens)
+        elif self.scheduler_type == "wsd":
```
Comment (dirkgr) [inline src/cookbook/model/builder.py:368]: Doesn't this also need a variant with and without warmup? Maybe better to factor out warmup or not?

**Review by dirkgr — CHANGES_REQUESTED** [review]
I don't have the full overview here, but I assume the scenarios you want are all annealing from a checkpoint. And then you want:
* without loading optimizer state, with LR warmup, to some arbitrary LR
* with loading optimizer state, with LR warmup, to some arbitrary LR
* with loading optimizer state, without LR warmup, starting from the peak LR the previous run finished at

Are they all covered?

File: src/cookbook/model/builder.py (hunk 1 above)
Comment (undfined) [reply src/cookbook/model/builder.py:104]: Yeah fair point. I can move them into separate methods no problem.

**Comment by undfined** [comment]
Closing in favor of: https://github.com/allenai/olmo-cookbook/pull/56
